src/custom-element.js

```diff
@@ -668,21 +701,57 @@ function createBaseAmpElementProto(win) {
         }
         this.implementation_.layout_ = this.layout_;
         this.implementation_.firstAttachedCallback();
-        if (!isUpgraded(this)) {
-          // amp:attached is dispatched from the ElementStub class when it replayed
-          // the firstAttachedCallback call.
+        if (!this.isUpgraded()) {
+          // amp:attached is dispatched from the ElementStub class when it
+          // replayed the firstAttachedCallback call.
           this.dispatchCustomEvent('amp:stubbed');
         } else {
           this.dispatchCustomEvent('amp:attached');
         }
       } catch (e) {
         reportError(e, this);
       }
+
+      // It's important to have this flag set in the end to avoid
+      // `resources.add` called twice if upgrade happens immediately.
+      this.everAttached = true;
     }
     this.resources_.add(this);
   };
 
   /**
+   * Try to upgrade the element with the provided implementation.
+   * @param {!./base-element.BaseElement} impl
+   * @package @final @this {!Element}
+   */
+  ElementProto.tryUpgrade = function(impl) {
```

Comment:
Since we have tests elsewhere `@package` doesn't actually work :)

`@private` is better because it can be obfuscated and devirtualized.